### PR TITLE
Feature/sync

### PR DESCRIPTION
**Pull request by GavanWilhite**

### CODE REVIEW

File: dist/aframe-altspace-component.js [hunk 1]
```diff
@@ -465,86 +467,744 @@
...
+						component.sceneEl.emit('clientjoined', {id: joinedClientId}, false);
+					}, 1);
+				});
+
+				this.clientsRef.on('child_removed', function(childSnapshot) {
+					var leftClientId = childSnapshot.val();
+					//let the master client flag get set first
+					setTimeout(function(){
+						component.sceneEl.emit('clientleft', {id: leftClientId}, false);
+					}, 1);
```
Comment (brianpeiris) [inline dist/aframe-altspace-component.js]: You can avoid these setTimeouts by registering the `child_added` and `child_removed` callbacks _before_ you register the `value` callback. If you reverse the order, `value` always gets called-back after child_added and child_removed http://stackoverflow.com/questions/33516346/why-does-the-firebase-event-child-added-fire-after-the-value-event

File: dist/aframe-altspace-component.js [hunk 2]
```diff
@@ -465,86 +467,744 @@
...
+	});
+
+	AFRAME.registerComponent('sync',
+	{
+		schema: {
+			mode: { default: 'link' }
+		},
+		init: function () {
+			var scene = document.querySelector('a-scene');
+			var syncSys = scene.systems['sync-system'];
```
Comment (brianpeiris) [inline dist/aframe-altspace-component.js:692]: If you rename `sync-system` to `sync`, a-frame will automatically associate the two and you'd be able to access the system from the component via `this.system` https://aframe.io/docs/0.3.0/core/systems.html#registering-a-system

File: dist/aframe-altspace-component.js [hunk 3]
```diff
@@ -465,86 +467,744 @@
...
+			var ownerRef;
+			var ownerId;
+			var isMine = false;
+
+			var component = this;
+
+			//Make sure someone always owns an object. If the owner leaves and we are the master client, we will take it.
+			//This ensures, for example, that synced animations keep playing
+			scene.addEventListener('clientleft', function(event){
+				var shouldTakeOwnership = (!ownerId || ownerId === event.detail.id) && syncSys.isMasterClient;
```
Comment (brianpeiris) [inline dist/aframe-altspace-component.js:706]: This reads strangely. What's with the `ownerId === event.detail.id` condition? It seems like it's saying "if I left and I'm the master, then take ownership", which doesn't make sense.

File: dist/aframe-altspace-component.js [hunk 4]
```diff
@@ -465,86 +467,744 @@
...
+
+				ownerRef.on('value',
+					function(snapshot) {
+						var newOwnerId = snapshot.val();
+
+						var gained = newOwnerId === syncSys.clientId && !isMine;
+						if (gained) component.el.emit('ownershipgained', null, false);
+
+
+						//note this also fires when we start up without ownership
```
Comment (brianpeiris) [inline dist/aframe-altspace-component.js:761]: You can avoid this if you check for `ownerId === undefined` right?

File: dist/aframe-altspace-component.js [hunk 5]
```diff
@@ -465,86 +467,744 @@
...
+					}
+				}
+			});
+			this.el.addEventListener('ownershipgained', function () {
+				for (var i = 0; i < stoppedAnimations.length; i++) {
+					var animation = stoppedAnimations[i];
+					animation.start();
+				}
+				stoppedAnimations = [];
+			});
```
Comment (brianpeiris) [inline dist/aframe-altspace-component.js:825]: Nice! Glad to see you could implement this cleanly.

Comment (brianpeiris) [reply dist/aframe-altspace-component.js:825]: Hmm, does this work for mixed-in animations?

File: dist/aframe-altspace-component.js [hunk 6]
```diff
@@ -465,86 +467,744 @@
...
+
+			positionRef.on('value', function (snapshot) {
+				if (sync.isMine) return;
+
+				var value = snapshot.val();
+				if (!value) return;
+
+				component.el.setAttribute('position', {x: value.x, y: value.y, z: value.z});
```
Comment (brianpeiris) [inline dist/aframe-altspace-component.js]: Hmm, is it necessary to create a new anon object? Can't you just use `value` directly?

File: dist/aframe-altspace-component.js [hunk 7]
```diff
@@ -465,86 +467,744 @@
...
+				  clearTimeout(timeout);
+				  previous = 0;
+				  timeout = context = args = null;
+				};
+
+				return throttled;
+			  };
+
+
+			this.el.addEventListener('componentchanged', throttle(onComponentChanged, this.updateRate));
```
Comment (brianpeiris) [inline dist/aframe-altspace-component.js]: Do we really have to throttle?

File: dist/aframe-altspace-component.js [hunk 8]
```diff
@@ -465,86 +467,744 @@
...
+				if (name !== 'material') return;
+				if (refChangedLocked) return;
+
+				if (oldData.color !== newData.color) {
+					setTimeout(function() {
+							componentChangedLock = true;
+							colorRef.set(newData.color);
+							componentChangedLock = false;
+						},
+						1);
```
Comment (brianpeiris) [inline dist/aframe-altspace-component.js]: Feels like we should just stick to the ownership mechanism. This is a little surprising and setTimeouts often lead to confusion (e.g. debugging gives you mysterious detached stack traces)
Being explicit about ownership feels right.

File: dist/aframe-altspace-component.js (hunk 2 above)
Comment (GavanWilhite) [reply dist/aframe-altspace-component.js:692]: Unfortunately this is broken at the moment (update gets double called with the system schema). Waiting on the A-Frame fix.

File: dist/aframe-altspace-component.js (hunk 3 above)
Comment (GavanWilhite) [reply dist/aframe-altspace-component.js:706]: That !ownerId part isn't necessary and I removed it. OwnerId is the current owner. So it's saying "if the owner left, and I'm the master client, then take ownership"

File: dist/aframe-altspace-component.js (hunk 4 above)
Comment (GavanWilhite) [reply dist/aframe-altspace-component.js:761]: Yeah.. not sure what the intended behavior should be. It makes the transform sync code simpler if it does the above behavior.

File: dist/aframe-altspace-component.js (hunk 5 above)
Comment (GavanWilhite) [reply dist/aframe-altspace-component.js:825]: Can you do that?

File: dist/aframe-altspace-component.js (hunk 1 above)
Comment (GavanWilhite) [reply dist/aframe-altspace-component.js]: Are you sure? I see a quote from the old docs saying "Value events are always triggered last"

File: dist/aframe-altspace-component.js (hunk 6 above)
Comment (GavanWilhite) [reply dist/aframe-altspace-component.js]: Huh, neat trick :-)

File: dist/aframe-altspace-component.js (hunk 2 above)
Comment (brianpeiris) [reply dist/aframe-altspace-component.js:692]: aww, ok

File: dist/aframe-altspace-component.js (hunk 3 above)
Comment (brianpeiris) [reply dist/aframe-altspace-component.js:706]: Ah, I misunderstood :+1:

File: dist/aframe-altspace-component.js (hunk 8 above)
Comment (GavanWilhite) [reply dist/aframe-altspace-component.js]: Yeah I think you're right. I'm not sure if that gets around the timeout, I'll have to check.

File: dist/aframe-altspace-component.js (hunk 7 above)
Comment (GavanWilhite) [reply dist/aframe-altspace-component.js]: Given how fast firebase fell over last time, I think so :P

File: dist/aframe-altspace-component.js (hunk 4 above)
Comment (brianpeiris) [reply dist/aframe-altspace-component.js:761]: Oh ok. Not a huge deal. Anything devs do in `ownershiplost` callbacks should be idempotent anyway.

File: dist/aframe-altspace-component.js (hunk 5 above)
Comment (brianpeiris) [reply dist/aframe-altspace-component.js:825]: Oh, I guess you can't. Nevermind

File: dist/aframe-altspace-component.js (hunk 7 above)
Comment (brianpeiris) [reply dist/aframe-altspace-component.js]: Well, that was an N^2 thing. But, ok. I guess it's fine if it doesn't visually affect animations too much. We'd probably want to add updateRate to the component schema though.

File: dist/aframe-altspace-component.js (hunk 1 above)
Comment (brianpeiris) [reply dist/aframe-altspace-component.js]: Odd, you're right it's better to take firebase at its word. Timeouts are still annoying, but I can't really think of a better approach right now.

Comment (brianpeiris) [inline dist/aframe-altspace-component.js]: Odd, you're right it's better to take firebase at its word. Timeouts are still annoying, but I can't really think of a better approach right now.

**Pull request merged**